paho/client.go

```diff
@@ -791,79 +812,154 @@ func (c *Client) Publish(ctx context.Context, p *Publish) (*PublishResponse, err
 
 	pb := p.Packet()
 
-	switch p.QoS {
-	case 0:
+	if p.QoS == 0 {
 		c.debug.Println("sending QoS0 message")
-		if _, err := pb.WriteTo(c.Conn); err != nil {
-			return nil, err
-		}
-		return nil, nil
-	case 1, 2:
-		return c.publishQoS12(ctx, pb)
-	}
-
-	return nil, fmt.Errorf("QoS isn't 0, 1 or 2")
-}
-
-func (c *Client) publishQoS12(ctx context.Context, pb *packets.Publish) (*PublishResponse, error) {
-	c.debug.Println("sending QoS12 message")
-	pubCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
-	defer cf()
-	if err := c.serverInflight.Acquire(pubCtx, 1); err != nil {
+		_, err := pb.WriteTo(c.Conn)
 		return nil, err
 	}
-	defer c.serverInflight.Release(1)
-	cpCtx := &CPContext{pubCtx, make(chan packets.ControlPacket, 1)}
-
-	mid, err := c.MIDs.Request(cpCtx)
-	if err != nil {
-		return nil, err
+	if p.QoS != 1 && p.QoS != 2 {
+		return nil, fmt.Errorf("QoS isn't 0, 1 or 2")
 	}
-	defer c.MIDs.Free(mid)
-	pb.PacketID = mid
 
-	if _, err := pb.WriteTo(c.Conn); err != nil {
+	resp, err := c.publishQoS12(pb)
+	if err != nil {
 		return nil, err
 	}
-	var resp packets.ControlPacket
 
 	select {
 	case <-pubCtx.Done():
-		if ctxErr := pubCtx.Err(); ctxErr != nil {
-			c.debug.Println(fmt.Sprintf("terminated due to context: %v", ctxErr))
-			return nil, ctxErr
+		c.OrphanedOps.AddOp(resp)
+		return nil, pubCtx.Err()
+	case r := <-resp:
+		switch r.Type {
+		case packets.PUBACK:
+			return PublishResponseFromPuback(r.Content.(*packets.Puback)), nil
+		case packets.PUBREC:
+			return PublishResponseFromPubrec(r.Content.(*packets.Pubrec)), nil
+		case packets.PUBCOMP:
+			return PublishResponseFromPubcomp(r.Content.(*packets.Pubcomp)), nil
+		default:
+			return nil, fmt.Errorf("unexpected response packet type %d", r.Type)
 		}
-	case resp = <-cpCtx.Return:
 	}
+}
 
-	switch pb.QoS {
-	case 1:
-		if resp.Type != packets.PUBACK {
-			return nil, fmt.Errorf("received %d instead of PUBACK", resp.Type)
+func (c *Client) publishQoS12(pb *packets.Publish) (<-chan packets.ControlPacket, error) {
+	c.debug.Println("sending QoS 1 or 2 message")
+	resp := make(chan packets.ControlPacket, 1)
+
+	respInternal := make(chan packets.ControlPacket)
+
+	var err error
+	if pb.PacketID == 0 {
+		pb.PacketID, err = c.MIDs.Request(&CPContext{context.TODO(), respInternal})
+	} else {
+		// ClaimID() is not yet implemented
+		err = fmt.Errorf("publishing with a specific PacketID is not yet implemented")
```

Comment:
This is here as a placeholder, but this is intended to handle the case where we are re-delivering a PUBLISH when resuming an existing session